### PR TITLE
constraints: add cpu cgroups documentation

### LINKED ISSUE #386: Add documentation about cpu cgroups
explain how kata implements cpu cgroups

### PR DESCRIPTION
**Pull request by devimc**
Explain cpu cgroups are implemented in kata containers

fixes #386

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: constraints/cpu.md [hunk 1]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+
+Kata Containers run over two layers of cgroups, the first layer is in the guest where
+only the workload is placed, the second layer is in the host that is more complex and
+might contain more than one process and task (thread) depending of the number of
+containers per POD and vCPUs per container. Following diagram represents a nginx container
+created with docker with the default number of vcpus.
+
+
+```
+docker run -dt --runtime=kata-runtime nginx
```
Comment (jodh-intel) [inline constraints/cpu.md]: Could you add the `$ ` shell prompt prefix.

File: constraints/cpu.md [hunk 2]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+       | nginx |
+    .--'-------'---.  .------------.
+    | Guest Cgroup |  | Kata agent |
+  .-'--------------'--'------------'.    .-----------.
+  |  Thread: Hypervisor's vCPU 0    |    | Kata Shim |
+ .'---------------------------------'.  .'-----------'.
+ |             Tasks                 |  |  Processes  |
+.'-----------------------------------'--'-------------'.
+|                    Host Cgroup                       |
+'------------------------------------------------------'
```
Comment (jodh-intel) [inline constraints/cpu.md]: ASCII ftw! Love it! :smile:

File: constraints/cpu.md [hunk 3]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+```
+
+In next sections is explained what is the difference between processes and tasks and why
```
Comment (jodh-intel) [inline constraints/cpu.md]: I'd simplify to just:

> The next sections explain the difference between processes and tasks and why only hypervisor vCPUs are constrained.

File: constraints/cpu.md [hunk 4]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+only hypervisor's vCPUs are constrained.
+
+### cpu cgroups in the guess
+
+Only the workload process including all its threads are placed into cpu cgroups, this means
+that kata agent and systemd run without constraints in the guest.
```
Comment (jodh-intel) [inline constraints/cpu.md]: "Kata agent" or `kata-agent`?

File: constraints/cpu.md [hunk 5]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+
+#### CPU pinning (cpuset cgroup)
+
+Kata Containers try to apply and honor the cgroups but sometimes that's not possible, this
```
Comment (jodh-intel) [inline constraints/cpu.md]: s/try/tries/. I'd split this into two sentences for easier reading. Something like:

> Kata Containers tries to apply and honor the cgroups but sometimes that is not possible. An example of this occurs with cpu cgroups when the number of virtual CPUs (in the guest) does not match the actual number of physical host CPUs.

File: constraints/cpu.md [hunk 6]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+is the case of cpuset cgroup where the number of virtual CPUs (guest) does not match with
+the actual number of physical CPUs (host).
+In Kata Containers to have a good performance and small memory footprint, the resources are
+hot added when they are needed, therefore the number of virtual resources is not the same
+than the number of physical resources. The problem with this approach is that it's not possible
```
Comment (jodh-intel) [inline constraints/cpu.md]: s/than/as/

File: constraints/cpu.md [hunk 7]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+to pin a process on a specific resource that is not present in the guest. To deal with this
+limitation and do not fail when the container is being created, Kata Containers do not apply
```
Comment (jodh-intel) [inline constraints/cpu.md]: - s/do not fail/to not fail/
- s/do not apply/does not apply/

File: constraints/cpu.md [hunk 8]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+the constraint in the first layer (guest) if the resource does not exist in the guest, but it
```
Comment (jodh-intel) [inline constraints/cpu.md]: s/does not exist/do not exist/

File: constraints/cpu.md [hunk 9]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+is applied in the second layer (host) where the hypervisor is running. In next sections is
```
Comment (jodh-intel) [inline constraints/cpu.md]: I'd simplify to:

> The next sections provide further details on what parts of the hypervisor are constrained.

File: constraints/cpu.md [hunk 10]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+explained with more details what parts of the hypervisor are constrained.
+
+### cpu cgroups in the host
+
+In Kata Containers the workloads run in a virtual machine that is managed and represented by a
+hypervisor running in the host, like other processes the hypervisor might use threads to realize
```
Comment (jodh-intel) [inline constraints/cpu.md]: I think we need a period somewhere around here. Maybe:

> ... hypervisor running in the host.

File: constraints/cpu.md [hunk 11]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+several tasks, for example IO and Network operations, one of the most important uses for the
+threads is as vCPUs, the processes running in the guest see these vCPUs as physical CPUs, while
+in the host those vCPU are just threads that are part of a process, this is the key to ensure
+workloads consumes only the amount of CPU resources that were assigned to it without impacting
+other operations. From user perspective the easier approach to implement it would be to take the
```
Comment (jodh-intel) [inline constraints/cpu.md]: Wow, the sentence ends here :) I'd definitely break this up into a number of smaller, more easily digestible sentences.

File: constraints/cpu.md [hunk 12]
```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
...
+whole hypervisor including its threads and move them into the cgroup, unfortunately this will
+impact negatively the performance, since vCPUs, IO and Network threads will be fighting for
+resources. Following table shows a random read performance comparison between a Kata Container
+with all its hypervisor threads in the cgroup and other with only its hypervisor vCPU threads,
+the difference is huge.
+
+
+| Bandwidth     | All threads   | vCPU threads | Units |
```
Comment (jodh-intel) [inline constraints/cpu.md]: It's a pain that gfm doesn't support column spans (for "Units") isn't it? :)

Comment (devimc) [reply constraints/cpu.md]: yeah it's  :smile:

File: constraints/cpu.md (hunk 8 above)
Comment (devimc) [reply constraints/cpu.md]: probably I'm wrong, but `the resource` is an `it` and it's singular, isn't it?

Comment (jodh-intel) [reply constraints/cpu.md]: Oops! Good spot - my mistake. Ignore that comment ;)

**Comment by devimc** [comment]
@grahamwhaley @jodh-intel  changes applied, thanks

**Pull request merged**